src/java/org/apache/cassandra/io/tries/TrieNode.java

```diff
@@ -0,0 +1,968 @@
+        }
+
+        @Override
+        public long greaterTransition(ByteBuffer src, int position, long positionLong, int searchIndex, long defaultValue)
+        {
+            if (searchIndex < 0)
+                searchIndex = -1 - searchIndex;
+            else
+                ++searchIndex;
+            int len = transitionRange(src, position);
+            for (; searchIndex < len; ++searchIndex)
+            {
+                long t = transition(src, position, positionLong, searchIndex);
+                if (t != -1)
+                    return t;
+            }
+            return defaultValue;
+        }
+
+        @Override
+        public long lesserTransition(ByteBuffer src, int position, long positionLong, int searchIndex, long defaultValue)
+        {
+            if (searchIndex == 0 || searchIndex == -1)
+                return defaultValue;
+
+            if (searchIndex < 0)
+                searchIndex = -2 - searchIndex;
+            else
+                --searchIndex;
+            for (; searchIndex >= 0; --searchIndex)
+            {
+                long t = transition(src, position, positionLong, searchIndex);
+                if (t != -1)
+                    return t;
+            }
+            assert false : "transition must always exist at 0 and we should not be called for less of that";
+            return defaultValue;
+        }
+
+        @Override
+        public int transitionByte(ByteBuffer src, int position, int childIndex)
+        {
+            if (childIndex >= transitionRange(src, position))
+                return Integer.MAX_VALUE;
+            int l = src.get(position + 1) & 0xFF;
+            return l + childIndex;
+        }
+
+        @Override
+        public int sizeofNode(SerializationNode<?> node)
+        {
+            int l = node.transition(0);
+            int r = node.transition(node.childCount() - 1);
+            return 3 + (r - l + 1) * bytesPerPointer;
+        }
+
+        @Override
+        public void serialize(DataOutput dest, SerializationNode<?> node, int payloadBits, long nodePosition) throws IOException
+        {
+            int childCount = node.childCount();
+            dest.writeByte((ordinal << 4) + (payloadBits & 0x0F));
+            int l = node.transition(0);
+            int r = node.transition(childCount - 1);
+            assert 0 <= l && l <= r && r <= 255;
+            dest.writeByte(l);
+            dest.writeByte(r - l);      // r is included, i.e. this is len - 1
+
+            for (int i = 0; i < childCount; ++i)
+            {
+                int next = node.transition(i);
+                while (l < next)
+                {
+                    writeBytes(dest, NULL_VALUE);
+                    ++l;
+                }
+                writeBytes(dest, -node.serializedPositionDelta(i, nodePosition));
+                ++l;
+            }
+        }
+    };
+
+    static final Dense12 DENSE_12 = new Dense12();
```

Comment:
nit: Should probably move these into `Dense12` itself to avoid potential classloading deadlocks